stan/math/rev/fun/trace_gen_quad_form.hpp

```diff
@@ -114,6 +114,179 @@ inline var trace_gen_quad_form(const Td& D, const Ta& A, const Tb& B) {
                                              Tb_scal, Rb, Cb>(baseVari));
 }
 
+/**
+ * Return the trace of D times the quadratic form of B and A.
+ * That is, `trace_gen_quad_form(D, A, B) = trace(D * B' * A * B).`
+ *
+ * This overload requires one of D, A, or B to be a `var_value<T>`
+ * where `T` inherits from EigenBase
+ *
+ * @tparam TD type of first matrix argument
+ * @tparam TA type of second matrix argument
+ * @tparam TB type of third matrix argument
+ *
+ * @param D multiplier
+ * @param A outside term in quadratic form
+ * @param B inner term in quadratic form
+ * @return trace(D * B' * A * B)
+ * @throw std::domain_error if A or D is not square
+ * @throw std::domain_error if A cannot be multiplied by B or B cannot
+ * be multiplied by D.
+ */
+template <typename Td, typename Ta, typename Tb,
+          require_any_var_matrix_t<Td, Ta, Tb>* = nullptr,
+          require_all_matrix_t<Td, Ta, Tb>* = nullptr>
+inline var trace_gen_quad_form(const Td& D, const Ta& A, const Tb& B) {
+  check_square("trace_gen_quad_form", "A", A);
+  check_square("trace_gen_quad_form", "D", D);
+  check_multiplicable("trace_gen_quad_form", "A", A, "B", B);
+  check_multiplicable("trace_gen_quad_form", "B", B, "D", D);
+
+  if (!is_constant<Ta>::value && !is_constant<Tb>::value
+      && !is_constant<Td>::value) {
+    arena_t<promote_scalar_t<var, Td>> arena_D = D;
+    arena_t<promote_scalar_t<var, Ta>> arena_A = A;
+    arena_t<promote_scalar_t<var, Tb>> arena_B = B;
+
+    var res = (arena_D.val_op() * arena_B.val_op().transpose()
+               * arena_A.val_op() * arena_B.val_op())
+                  .trace();
+
+    reverse_pass_callback([arena_A, arena_B, arena_D, res]() mutable {
+      double C_adj = res.adj();
+
+      auto BDT = (arena_B.val_op() * arena_D.val_op().transpose()).eval();
+      auto AB = (arena_A.val_op() * arena_B.val_op()).eval();
```

Comment:
[optional] You compute `A.val() * B.val()` in the forward pass you could just store it in the forward pass and reuse it

```suggestion
   arena_t<decltype(arena_A.val_op() * arena_B.val_op())> ab_val_mul = arena_A.val_op() * arena_B.val_op()
    var res = (arena_D.val_op() * arena_B.val_op().transpose() * ab_val_mul).trace();

    reverse_pass_callback([arena_A, arena_B, arena_D, res]() mutable {
      double C_adj = res.adj();

      auto BDT = (arena_B.val_op() * arena_D.val_op().transpose()).eval();
```

Reply:
Actually is it true if `B` is a square matrix that `D * B.transpose() = (B * D.transpose()).transpose()`? If so I think you can reuse that as well down here

https://godbolt.org/z/5ohjK5

Reply:
If so then you can use the above in the other specializations as well

Reply:
Woof, good catch, yes can rework this.

Reply:
I updated this to use the AB/BDT temporaries in the forward pass.